File: server/src/routes/health.ts
import { Hono } from 'hono';
import { getDb } from '../db/index.js';
import { checkKeyHealth, checkAllKeys } from '../services/health.js';
import { hasProvider } from '../providers/index.js';
import type { Platform } from '@freellmapi/shared/types.js';
import * as schema from '../db/schema.js';
import { sql, eq, desc, asc } from 'drizzle-orm';

export const healthRouter = new Hono();

// Get health status for all platforms
healthRouter.get('/', async (c) => {
  const db = getDb();

  const platforms = db.select({
    platform: schema.apiKeys.platform,
    total_keys: sql<number>`COUNT(*)`,
    healthy_keys: sql<number>`SUM(CASE WHEN ${schema.apiKeys.status} = 'healthy' THEN 1 ELSE 0 END)`,
    rate_limited_keys: sql<number>`SUM(CASE WHEN ${schema.apiKeys.status} = 'rate_limited' THEN 1 ELSE 0 END)`,
    invalid_keys: sql<number>`SUM(CASE WHEN ${schema.apiKeys.status} = 'invalid' THEN 1 ELSE 0 END)`,
    error_keys: sql<number>`SUM(CASE WHEN ${schema.apiKeys.status} = 'error' THEN 1 ELSE 0 END)`,
    unknown_keys: sql<number>`SUM(CASE WHEN ${schema.apiKeys.status} = 'unknown' THEN 1 ELSE 0 END)`,
    enabled_keys: sql<number>`SUM(CASE WHEN ${schema.apiKeys.enabled} = 1 THEN 1 ELSE 0 END)`
  })
  .from(schema.apiKeys)
  .groupBy(schema.apiKeys.platform)
  .all();

  const keys = db.select({
    id: schema.apiKeys.id,
    platform: schema.apiKeys.platform,
    label: schema.apiKeys.label,
    status: schema.apiKeys.status,
    enabled: schema.apiKeys.enabled,
    createdAt: schema.apiKeys.createdAt,
    lastCheckedAt: schema.apiKeys.lastCheckedAt
  })
  .from(schema.apiKeys)
  .orderBy(schema.apiKeys.platform, desc(schema.apiKeys.createdAt))
  .all();

   return c.json({
     platforms: platforms.map(p => ({
       platform: p.platform as Platform,
       hasProvider: hasProvider(p.platform as Platform),
       totalKeys: p.total_keys,
       healthyKeys: p.healthy_keys ?? 0,
       rateLimitedKeys: p.rate_limited_keys ?? 0,
       invalidKeys: p.invalid_keys ?? 0,
       errorKeys: p.error_keys ?? 0,
       unknownKeys: p.unknown_keys ?? 0,
       enabledKeys: p.enabled_keys ?? 0,
     })),
     keys: keys.map(k => ({
       id: k.id,
       platform: k.platform as Platform,
       label: k.label,
       status: k.status,
       enabled: k.enabled === 1,
       createdAt: k.createdAt,
       lastCheckedAt: k.lastCheckedAt,
     })),
   });
});

// Check a specific key
healthRouter.post('/check/:keyId', async (c) => {
  const keyId = parseInt(c.req.param('keyId'), 10);
  if (isNaN(keyId)) {
    c.status(400)
    return c.json({ error: { message: 'Invalid key ID' } });
  }

  const status = await checkKeyHealth(keyId);
  return c.json({ keyId, status });
});

// Check all keys
healthRouter.post('/check-all', async (c) => {
  await checkAllKeys();
  c.status(200)
  return c.json({ success: true });
});

File: crates/server/src/routes/health.rs
//! Port of `server/src/routes/health.ts` — per-platform health summary listing
//! and the manual key-check endpoints.

use axum::extract::Path;
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::routing::get;
use axum::Json;
use serde_json::{json, Value};

use crate::db::connection::db;
use crate::db::schema::{ApiKeyRow, API_KEY_COLS};
use crate::providers::has_provider;
use crate::services::health::{check_all_keys, check_key_health};

type PlatformHealthRow = (String, i64, i64, i64, i64, i64, i64, i64);

/// `parseInt(param, 10)`-style id parse; `None` for NaN.
fn parse_id(s: &str) -> Option<i64> {
    s.parse::<i64>().ok()
}

fn json_error(status: u16, message: &str) -> Response {
    (
        StatusCode::from_u16(status).unwrap_or(StatusCode::BAD_REQUEST),
        Json(json!({ "error": { "message": message } })),
    )
        .into_response()
}

/// `GET /` — per-platform key-status aggregation plus the full key list.
async fn get_health_status() -> Response {
    let (platform_rows, keys): (
        Vec<PlatformHealthRow>,
        Vec<ApiKeyRow>,
    ) = {
        let conn = db().lock().await;

        let platforms_sql = {
            "SELECT platform, COUNT(*), \
                    SUM(CASE WHEN status = 'healthy' THEN 1 ELSE 0 END), \
                    SUM(CASE WHEN status = 'rate_limited' THEN 1 ELSE 0 END), \
                    SUM(CASE WHEN status = 'invalid' THEN 1 ELSE 0 END), \
                    SUM(CASE WHEN status = 'error' THEN 1 ELSE 0 END), \
                    SUM(CASE WHEN status = 'unknown' THEN 1 ELSE 0 END), \
                    SUM(CASE WHEN enabled = 1 THEN 1 ELSE 0 END) \
             FROM api_keys GROUP BY platform"
        };
        let aggs: rusqlite::Result<Vec<PlatformHealthRow>> =
            conn.prepare(platforms_sql).and_then(|mut s| {
                s.query_map([], |r| {
                    Ok((
                        r.get::<_, String>(0)?,
                        r.get::<_, i64>(1)?,
                        r.get::<_, i64>(2)?,
                        r.get::<_, i64>(3)?,
                        r.get::<_, i64>(4)?,
                        r.get::<_, i64>(5)?,
                        r.get::<_, i64>(6)?,
                        r.get::<_, i64>(7)?,
                    ))
                })?
                .collect()
            });

        let keys_sql = format!(
            "SELECT {API_KEY_COLS} FROM api_keys ORDER BY platform, created_at DESC"
        );
        let keys: rusqlite::Result<Vec<ApiKeyRow>> = conn
            .prepare(&keys_sql)
            .and_then(|mut s| {
                s.query_map([], ApiKeyRow::from_row)?.collect::<rusqlite::Result<Vec<_>>>()
            });

        match (aggs, keys) {
            (Ok(a), Ok(k)) => (a, k),
            (Err(e), _) | (_, Err(e)) => return crate::app::error_text(500, &e.to_string()),
        }
    };

    // `hasProvider` locks the DB internally, so run it after the guard above is dropped.
    let mut platforms_json = Vec::with_capacity(platform_rows.len());
    for (platform, total_keys, healthy, rate_limited, invalid, error, unknown, enabled_keys) in
        platform_rows
    {
        let has = has_provider(&platform).await;
        platforms_json.push(json!({
            "platform": platform,
            "hasProvider": has,
            "totalKeys": total_keys,
            "healthyKeys": healthy,
            "rateLimitedKeys": rate_limited,
            "invalidKeys": invalid,
            "errorKeys": error,
            "unknownKeys": unknown,
            "enabledKeys": enabled_keys,
        }));
    }

    let keys_json: Vec<Value> = keys
        .iter()
        .map(|k| {
            json!({
                "id": k.id,
                "platform": k.platform,
                "label": k.label,
                "status": k.status,
                "enabled": k.enabled != 0,
                "createdAt": k.created_at,
                "lastCheckedAt": k.last_checked_at,
            })
        })
        .collect();

    Json(json!({ "platforms": platforms_json, "keys": keys_json })).into_response()
}

/// `POST /check/:keyId`
async fn check_key(Path(key_id): Path<String>) -> Response {
    let Some(key_id) = parse_id(&key_id) else {
        return json_error(400, "Invalid key ID");
    };
    let status = check_key_health(key_id).await;
    Json(json!({ "keyId": key_id, "status": status })).into_response()
}

/// `POST /check-all`
async fn check_all() -> Response {
    check_all_keys().await;
    Json(json!({ "success": true })).into_response()
}

pub fn router() -> axum::Router {
    axum::Router::new()
        .route("/", get(get_health_status))
        .route("/check/{keyId}", axum::routing::post(check_key))
        .route("/check-all", axum::routing::post(check_all))
}
